// @ngageoint/simple-features-js/dist/lib/util/centroid/CentroidSurface.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CentroidSurface = void 0;
var internal_1 = require("../../internal");
/**
 * Calculate the centroid from surface based geometries. Implementation based on
 * the JTS (Java Topology Suite) CentroidArea.
 */
var CentroidSurface = /** @class */ (function () {
    /**
     * Constructor
     */
    function CentroidSurface() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        /**
         * Area sum
         */
        this._area = 0;
        /**
         * Sum of surface point locations
         */
        this._sum = new internal_1.Point();
        if (args.length === 1 && args[0] instanceof internal_1.Geometry) {
            this.add(args[0]);
        }
    }
    /**
     * Add a surface based dimension 2 geometry to the centroid total. Ignores
     * dimension 0 and 1 geometries.
     *
     * @param geometry
     *            geometry
     */
    CentroidSurface.prototype.add = function (geometry) {
        var e_1, _a;
        var geometryType = geometry.geometryType;
        switch (geometryType) {
            case internal_1.GeometryType.POLYGON:
            case internal_1.GeometryType.TRIANGLE:
                this.addPolygon(geometry);
                break;
            case internal_1.GeometryType.MULTIPOLYGON:
                this.addPolygons(geometry.polygons);
                break;
            case internal_1.GeometryType.CURVEPOLYGON:
                this.addCurvePolygon(geometry);
                break;
            case internal_1.GeometryType.POLYHEDRALSURFACE:
            case internal_1.GeometryType.TIN:
                this.addPolygons(geometry.polygons);
                break;
            case internal_1.GeometryType.GEOMETRYCOLLECTION:
            case internal_1.GeometryType.MULTICURVE:
            case internal_1.GeometryType.MULTISURFACE:
                try {
                    for (var _b = __values(geometry.geometries), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var subGeometry = _c.value;
                        this.add(subGeometry);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                break;
            case internal_1.GeometryType.POINT:
            case internal_1.GeometryType.MULTIPOINT:
            case internal_1.GeometryType.LINESTRING:
            case internal_1.GeometryType.CIRCULARSTRING:
            case internal_1.GeometryType.MULTILINESTRING:
            case internal_1.GeometryType.COMPOUNDCURVE:
                // Doesn't contribute to surface dimension
                break;
            default:
                throw new internal_1.SFException("Unsupported CendroidSurface Geometry Type: " + geometryType);
        }
    };
    /**
     * Add polygons to the centroid total
     *
     * @param polygons polygons
     */
    CentroidSurface.prototype.addPolygons = function (polygons) {
        var e_2, _a;
        try {
            for (var polygons_1 = __values(polygons), polygons_1_1 = polygons_1.next(); !polygons_1_1.done; polygons_1_1 = polygons_1.next()) {
                var polygon = polygons_1_1.value;
                this.addPolygon(polygon);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (polygons_1_1 && !polygons_1_1.done && (_a = polygons_1.return)) _a.call(polygons_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    /**
     * Add a polygon to the centroid total
     *
     * @param polygon
     *            polygon
     */
    CentroidSurface.prototype.addPolygon = function (polygon) {
        var rings = polygon.rings;
        this.addLineString(rings[0]);
        for (var i = 1; i < rings.length; i++) {
            this.addHole(rings[i]);
        }
    };
    /**
     * Add a curve polygon to the centroid total
     *
     * @param curvePolygon
     *            curve polygon
     */
    CentroidSurface.prototype.addCurvePolygon = function (curvePolygon) {
        var e_3, _a, e_4, _b;
        var rings = curvePolygon.rings;
        var curve = rings[0];
        var curveGeometryType = curve.geometryType;
        switch (curveGeometryType) {
            case internal_1.GeometryType.COMPOUNDCURVE:
                try {
                    for (var _c = __values(curve.lineStrings), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var lineString = _d.value;
                        this.addLineString(lineString);
                    }
                }
                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                    }
                    finally { if (e_3) throw e_3.error; }
                }
                break;
            case internal_1.GeometryType.LINESTRING:
            case internal_1.GeometryType.CIRCULARSTRING:
                this.addLineString(curve);
                break;
            default:
                throw new internal_1.SFException("Unexpected Curve Type: " + curveGeometryType);
        }
        for (var i = 1; i < rings.length; i++) {
            var curveHole = rings[i];
            var curveHoleGeometryType = curveHole.geometryType;
            switch (curveHoleGeometryType) {
                case internal_1.GeometryType.COMPOUNDCURVE:
                    try {
                        for (var _e = (e_4 = void 0, __values(curve.lineStrings)), _f = _e.next(); !_f.done; _f = _e.next()) {
                            var lineString = _f.value;
                            this.addHole(lineString);
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                    break;
                case internal_1.GeometryType.LINESTRING:
                case internal_1.GeometryType.CIRCULARSTRING:
                    this.addHole(curveHole);
                    break;
                default:
                    throw new internal_1.SFException("Unexpected Curve Type: " + curveHoleGeometryType);
            }
        }
    };
    /**
     * Add a line string to the centroid total
     * @param lineString line string
     */
    CentroidSurface.prototype.addLineString = function (lineString) {
        this.addOrSubtractLine(true, lineString);
    };
    /**
     * Add a line string hole to subtract from the centroid total
     * @param lineString line string
     */
    CentroidSurface.prototype.addHole = function (lineString) {
        this.addOrSubtractLine(false, lineString);
    };
    /**
     * Add or subtract a line string to or from the centroid total
     * @param positive true if an addition, false if a subtraction
     * @param lineString line string
     */
    CentroidSurface.prototype.addOrSubtractLine = function (positive, lineString) {
        var points = lineString.points;
        var firstPoint = points[0];
        if (this._base == null) {
            this._base = firstPoint;
        }
        for (var i = 0; i < points.length - 1; i++) {
            var point = points[i];
            var nextPoint = points[i + 1];
            this.addTriangle(positive, this._base, point, nextPoint);
        }
        var lastPoint = points[points.length - 1];
        if (firstPoint.x != lastPoint.x || firstPoint.y != lastPoint.y) {
            this.addTriangle(positive, this._base, lastPoint, firstPoint);
        }
    };
    /**
     * Add or subtract a triangle of points to or from the centroid total
     * @param positive true if an addition, false if a subtraction
     * @param point1 point 1
     * @param point2 point 2
     * @param point3 point 3
     */
    CentroidSurface.prototype.addTriangle = function (positive, point1, point2, point3) {
        var sign = positive ? 1.0 : -1.0;
        var triangleCenter3 = CentroidSurface.centroid3(point1, point2, point3);
        var area2 = CentroidSurface.area2(point1, point2, point3);
        this._sum.x = (this._sum.x + (sign * area2 * triangleCenter3.x));
        this._sum.y = (this._sum.y + (sign * area2 * triangleCenter3.y));
        this._area += sign * area2;
    };
    /**
     * Calculate three times the centroid of the point triangle
     * @param point1 point 1
     * @param point2 point 2
     * @param point3 point 3
     * @return 3 times centroid point
     */
    CentroidSurface.centroid3 = function (point1, point2, point3) {
        var x = point1.x + point2.x + point3.x;
        var y = point1.y + point2.y + point3.y;
        return new internal_1.Point(x, y);
    };
    /**
     * Calculate twice the area of the point triangle
     *
     * @param point1
     *            point 1
     * @param point2
     *            point 2
     * @param point3
     *            point 3
     * @return 2 times triangle area
     */
    CentroidSurface.area2 = function (point1, point2, point3) {
        return (point2.x - point1.x)
            * (point3.y - point1.y)
            - (point3.x - point1.x)
                * (point2.y - point1.y);
    };
    /**
     * Get the centroid point
     * @return centroid point
     */
    CentroidSurface.prototype.getCentroid = function () {
        return new internal_1.Point((this._sum.x / 3.0) / this._area, (this._sum.y / 3.0) / this._area);
    };
    return CentroidSurface;
}());
exports.CentroidSurface = CentroidSurface;
