// @ngageoint/simple-features-js/dist/lib/util/centroid/CentroidCurve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CentroidCurve = void 0;
var internal_1 = require("../../internal");
/**
 * Calculate the centroid from curve based geometries. Implementation based on
 * the JTS (Java Topology Suite) CentroidLine.
 */
var CentroidCurve = /** @class */ (function () {
    /**
     * Constructor
     */
    function CentroidCurve() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        /**
         * Sum of curve point locations
         */
        this._sum = new internal_1.Point();
        /**
         * Total length of curves
         */
        this._totalLength = 0;
        if (args.length === 1 && args[0] instanceof internal_1.Geometry) {
            this.add(args[0]);
        }
    }
    /**
     * Add a curve based dimension 1 geometry to the centroid total. Ignores
     * dimension 0 geometries.
     * @param geometry geometry
     */
    CentroidCurve.prototype.add = function (geometry) {
        var _this = this;
        var geometryType = geometry.geometryType;
        switch (geometryType) {
            case internal_1.GeometryType.LINESTRING:
            case internal_1.GeometryType.CIRCULARSTRING:
                this.addLineString(geometry);
                break;
            case internal_1.GeometryType.MULTILINESTRING:
                this.addLineStrings(geometry.lineStrings);
                break;
            case internal_1.GeometryType.COMPOUNDCURVE:
                this.addLineStrings(geometry.lineStrings);
                break;
            case internal_1.GeometryType.GEOMETRYCOLLECTION:
            case internal_1.GeometryType.MULTICURVE:
            case internal_1.GeometryType.MULTISURFACE:
                geometry.geometries.forEach(function (geometry) {
                    _this.add(geometry);
                });
                break;
            case internal_1.GeometryType.POINT:
            case internal_1.GeometryType.MULTIPOINT:
                // Doesn't contribute to curve dimension
                break;
            default:
                throw new internal_1.SFException("Unsupported CentroidCurve Geometry Type: " + geometryType);
        }
    };
    /**
     * Add line strings to the centroid total
     * @param lineStrings line strings
     */
    CentroidCurve.prototype.addLineStrings = function (lineStrings) {
        var _this = this;
        lineStrings.forEach(function (lineString) { return _this.addLineString(lineString); });
    };
    /**
     * Add a line string to the centroid total
     * @param lineString line string
     */
    CentroidCurve.prototype.addLineString = function (lineString) {
        this.addPoints(lineString.points);
    };
    /**
     * Add points to the centroid total
     *
     * @param points
     *            points
     */
    CentroidCurve.prototype.addPoints = function (points) {
        for (var i = 0; i < points.length - 1; i++) {
            var point = points[i];
            var nextPoint = points[i + 1];
            var length_1 = internal_1.GeometryUtils.distance(point, nextPoint);
            this._totalLength += length_1;
            var midX = (point.x + nextPoint.x) / 2;
            this._sum.x = this._sum.x + (length_1 * midX);
            var midY = (point.y + nextPoint.y) / 2;
            this._sum.y = this._sum.y + (length_1 * midY);
        }
    };
    /**
     * Get the centroid point
     * @return centroid point
     */
    CentroidCurve.prototype.getCentroid = function () {
        return new internal_1.Point(this._sum.x / this._totalLength, this._sum.y / this._totalLength);
    };
    return CentroidCurve;
}());
exports.CentroidCurve = CentroidCurve;
