// @ngageoint/simple-features-js/dist/lib/util/centroid/DegreesCentroid.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DegreesCentroid = void 0;
var internal_1 = require("../../internal");
/**
 * Centroid calculations for geometries in degrees
 *
 */
var DegreesCentroid = /** @class */ (function () {
    /**
     * Constructor
     */
    function DegreesCentroid() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        /**
         * Number of points
         */
        this._points = 0;
        /**
         * x sum
         */
        this._x = 0;
        /**
         * y sum
         */
        this._y = 0;
        /**
         * z sum
         */
        this._z = 0;
        if (args.length === 1 && args[0] instanceof internal_1.Geometry) {
            this._geometry = args[0];
        }
    }
    /**
     * Get the degree geometry centroid
     * @param geometry geometry
     * @return centroid point
     */
    DegreesCentroid.getCentroid = function (geometry) {
        return new DegreesCentroid(geometry).getCentroid();
    };
    /**
     * Get the centroid point
     * @return centroid point
     */
    DegreesCentroid.prototype.getCentroid = function () {
        var centroid;
        if (this._geometry.geometryType == internal_1.GeometryType.POINT) {
            centroid = this._geometry;
        }
        else {
            this.calculate(this._geometry);
            this._x = this._x / this._points;
            this._y = this._y / this._points;
            this._z = this._z / this._points;
            var centroidLongitude = Math.atan2(this._y, this._x);
            var centroidLatitude = Math.atan2(this._z, Math.sqrt(this._x * this._x + this._y * this._y));
            centroid = new internal_1.Point(internal_1.GeometryUtils.radiansToDegrees(centroidLongitude), internal_1.GeometryUtils.radiansToDegrees(centroidLatitude));
        }
        return centroid;
    };
    /**
     * Add to the centroid calculation for the Geometry
     *
     * @param geometry
     *            Geometry
     */
    DegreesCentroid.prototype.calculate = function (geometry) {
        var geometryType = geometry.geometryType;
        switch (geometryType) {
            case internal_1.GeometryType.GEOMETRY:
                throw new internal_1.SFException("Unexpected Geometry Type of " + geometryType + " which is abstract");
            case internal_1.GeometryType.POINT:
                this.calculatePoint(geometry);
                break;
            case internal_1.GeometryType.LINESTRING:
            case internal_1.GeometryType.CIRCULARSTRING:
                this.calculateLineString(geometry);
                break;
            case internal_1.GeometryType.POLYGON:
            case internal_1.GeometryType.TRIANGLE:
                this.calculatePolygon(geometry);
                break;
            case internal_1.GeometryType.GEOMETRYCOLLECTION:
            case internal_1.GeometryType.MULTIPOINT:
            case internal_1.GeometryType.MULTICURVE:
            case internal_1.GeometryType.MULTILINESTRING:
            case internal_1.GeometryType.MULTISURFACE:
            case internal_1.GeometryType.MULTIPOLYGON:
                this.calculateGeometryCollection(geometry);
                break;
            case internal_1.GeometryType.COMPOUNDCURVE:
                this.calculateCompoundCurve(geometry);
                break;
            case internal_1.GeometryType.CURVEPOLYGON:
                this.calculateCurvePolygon(geometry);
                break;
            case internal_1.GeometryType.CURVE:
                throw new internal_1.SFException("Unexpected Geometry Type of "
                    + geometryType + " which is abstract");
            case internal_1.GeometryType.SURFACE:
                throw new internal_1.SFException("Unexpected Geometry Type of "
                    + geometryType + " which is abstract");
            case internal_1.GeometryType.POLYHEDRALSURFACE:
            case internal_1.GeometryType.TIN:
                this.calculatePolyhedralSurface(geometry);
                break;
            default:
                throw new internal_1.SFException("Geometry Type not supported: " + geometryType);
        }
    };
    /**
     * Add to the centroid calculation for the Point
     * @param point Point
     */
    DegreesCentroid.prototype.calculatePoint = function (point) {
        var latitude = internal_1.GeometryUtils.degreesToRadians(point.y);
        var longitude = internal_1.GeometryUtils.degreesToRadians(point.x);
        var cosLatitude = Math.cos(latitude);
        this._x += cosLatitude * Math.cos(longitude);
        this._y += cosLatitude * Math.sin(longitude);
        this._z += Math.sin(latitude);
        this._points++;
    };
    /**
     * Add to the centroid calculation for the Line String
     * @param lineString Line String
     */
    DegreesCentroid.prototype.calculateLineString = function (lineString) {
        var _this = this;
        lineString.points.forEach(function (point) { return _this.calculatePoint(point); });
    };
    /**
     * Add to the centroid calculation for the Polygon
     * @param polygon Polygon
     */
    DegreesCentroid.prototype.calculatePolygon = function (polygon) {
        if (polygon.numRings() > 0) {
            var exteriorRing = polygon.getExteriorRing();
            var numPoints = exteriorRing.numPoints();
            if (internal_1.GeometryUtils.closedPolygonRing(exteriorRing)) {
                numPoints--;
            }
            for (var i = 0; i < numPoints; i++) {
                this.calculatePoint(exteriorRing.getPoint(i));
            }
        }
    };
    /**
     * Add to the centroid calculation for the Geometry Collection
     * @param geometryCollection Geometry Collection
     */
    DegreesCentroid.prototype.calculateGeometryCollection = function (geometryCollection) {
        var e_1, _a;
        try {
            for (var _b = __values(geometryCollection.geometries), _c = _b.next(); !_c.done; _c = _b.next()) {
                var geometry = _c.value;
                this.calculate(geometry);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * Add to the centroid calculation for the Compound Curve
     * @param compoundCurve Compound Curve
     */
    DegreesCentroid.prototype.calculateCompoundCurve = function (compoundCurve) {
        var e_2, _a;
        try {
            for (var _b = __values(compoundCurve.lineStrings), _c = _b.next(); !_c.done; _c = _b.next()) {
                var lineString = _c.value;
                this.calculateLineString(lineString);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    /**
     * Add to the centroid calculation for the Curve Polygon
     * @param curvePolygon Curve Polygon
     */
    DegreesCentroid.prototype.calculateCurvePolygon = function (curvePolygon) {
        var e_3, _a;
        try {
            for (var _b = __values(curvePolygon.rings), _c = _b.next(); !_c.done; _c = _b.next()) {
                var ring = _c.value;
                this.calculate(ring);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
    };
    /**
     * Add to the centroid calculation for the Polyhedral Surface
     * @param polyhedralSurface Polyhedral Surface
     */
    DegreesCentroid.prototype.calculatePolyhedralSurface = function (polyhedralSurface) {
        var e_4, _a;
        try {
            for (var _b = __values(polyhedralSurface.polygons), _c = _b.next(); !_c.done; _c = _b.next()) {
                var polygon = _c.value;
                this.calculatePolygon(polygon);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
    };
    return DegreesCentroid;
}());
exports.DegreesCentroid = DegreesCentroid;
