// @ngageoint/simple-features-js/dist/lib/CompoundCurve.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompoundCurve = void 0;
var internal_1 = require("./internal");
/**
 * Compound Curve, Curve sub type
 *
 * @author osbornb
 */
var CompoundCurve = /** @class */ (function (_super) {
    __extends(CompoundCurve, _super);
    /**
     * Constructor
     */
    function CompoundCurve() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.COMPOUNDCURVE, false, false) || this;
            _this._lineStrings = [];
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.COMPOUNDCURVE, args[0], args[1]) || this;
            _this._lineStrings = [];
        }
        else if (args.length === 1 && args[0].length != null) {
            _this = _super.call(this, internal_1.GeometryType.COMPOUNDCURVE, internal_1.Geometry.hasZ(args[0]), internal_1.Geometry.hasM(args[0])) || this;
            _this._lineStrings = args[0] || [];
        }
        else if (args.length === 1 && args[0] instanceof internal_1.LineString) {
            _this = _super.call(this, internal_1.GeometryType.COMPOUNDCURVE, args[0].hasZ, args[0].hasM) || this;
            _this._lineStrings = [];
            _this.addLineString(args[0]);
        }
        else if (args.length === 1 && args[0] instanceof CompoundCurve) {
            _this = _super.call(this, internal_1.GeometryType.COMPOUNDCURVE, args[0].hasZ, args[0].hasM) || this;
            _this._lineStrings = [];
            args[0].lineStrings.forEach(function (lineString) { return _this.addLineString(lineString.copy()); });
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
            _this._lineStrings = [];
        }
        return _this;
    }
    Object.defineProperty(CompoundCurve.prototype, "lineStrings", {
        /**
         * Get the line strings
         * @return line strings
         */
        get: function () {
            return this._lineStrings;
        },
        /**
         * Set the line strings
         * @param lineStrings line strings
         */
        set: function (lineStrings) {
            this._lineStrings = lineStrings;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add a line string
     * @param lineString line string
     */
    CompoundCurve.prototype.addLineString = function (lineString) {
        this._lineStrings.push(lineString);
        this.updateZM(lineString);
    };
    /**
     * Add line strings
     * @param lineStrings line strings
     */
    CompoundCurve.prototype.addLineStrings = function (lineStrings) {
        var e_1, _a;
        try {
            for (var lineStrings_1 = __values(lineStrings), lineStrings_1_1 = lineStrings_1.next(); !lineStrings_1_1.done; lineStrings_1_1 = lineStrings_1.next()) {
                var lineString = lineStrings_1_1.value;
                this.addLineString(lineString);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (lineStrings_1_1 && !lineStrings_1_1.done && (_a = lineStrings_1.return)) _a.call(lineStrings_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * Get the number of line strings
     * @return number of line strings
     */
    CompoundCurve.prototype.numLineStrings = function () {
        return this._lineStrings.length;
    };
    /**
     * Returns the Nth line string
     * @param n  nth line string to return
     * @return line string
     */
    CompoundCurve.prototype.getLineString = function (n) {
        return this._lineStrings[n];
    };
    /**
     * {@inheritDoc}
     */
    CompoundCurve.prototype.startPoint = function () {
        var e_2, _a;
        var startPoint = null;
        if (!this.isEmpty()) {
            try {
                for (var _b = __values(this._lineStrings), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var lineString = _c.value;
                    if (!lineString.isEmpty()) {
                        startPoint = lineString.startPoint();
                        break;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        return startPoint;
    };
    /**
     * {@inheritDoc}
     */
    CompoundCurve.prototype.endPoint = function () {
        var endPoint = null;
        if (!this.isEmpty()) {
            for (var i = this._lineStrings.length - 1; i >= 0; i--) {
                var lineString = this._lineStrings[i];
                if (!lineString.isEmpty()) {
                    endPoint = lineString.endPoint();
                    break;
                }
            }
        }
        return endPoint;
    };
    /**
     * {@inheritDoc}
     */
    CompoundCurve.prototype.isSimple = function () {
        return internal_1.ShamosHoey.simplePolygonLineStrings(this._lineStrings);
    };
    /**
     * {@inheritDoc}
     */
    CompoundCurve.prototype.copy = function () {
        return new CompoundCurve(this);
    };
    /**
     * {@inheritDoc}
     */
    CompoundCurve.prototype.isEmpty = function () {
        return this._lineStrings.length === 0;
    };
    /**
     * {@inheritDoc}
     */
    CompoundCurve.prototype.equals = function (obj) {
        var equal = true;
        if (obj instanceof CompoundCurve && this.numLineStrings() === obj.numLineStrings()) {
            for (var i = 0; i < this._lineStrings.length; i++) {
                if (!this.getLineString(i).equals(obj.getLineString(i))) {
                    equal = false;
                    break;
                }
            }
        }
        else {
            equal = false;
        }
        return equal;
    };
    return CompoundCurve;
}(internal_1.Curve));
exports.CompoundCurve = CompoundCurve;
