// @ngageoint/simple-features-js/dist/lib/CurvePolygon.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurvePolygon = void 0;
var internal_1 = require("./internal");
/**
 * A planar surface defined by an exterior ring and zero or more interior ring.
 * Each ring is defined by a Curve instance.
 * @param <T> curve type
 */
var CurvePolygon = /** @class */ (function (_super) {
    __extends(CurvePolygon, _super);
    /**
     * Constructor
     */
    function CurvePolygon() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.CURVEPOLYGON, false, false) || this;
            _this._rings = [];
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.CURVEPOLYGON, args[0], args[1]) || this;
            _this._rings = [];
        }
        else if (args.length === 1 && args[0].length != null) {
            _this = _super.call(this, internal_1.GeometryType.CURVEPOLYGON, internal_1.Geometry.hasZ(args[0]), internal_1.Geometry.hasM(args[0])) || this;
            _this.rings = args[0] || [];
        }
        else if (args.length === 1 && args[0] instanceof internal_1.Curve) {
            _this = _super.call(this, internal_1.GeometryType.CURVEPOLYGON, args[0].hasZ, args[0].hasM) || this;
            _this._rings = [];
            _this.addRing(args[0]);
        }
        else if (args.length === 1 && args[0] instanceof CurvePolygon) {
            _this = _super.call(this, internal_1.GeometryType.CURVEPOLYGON, args[0].hasZ, args[0].hasM) || this;
            _this._rings = [];
            args[0].rings.forEach(function (ring) { return _this.addRing(ring.copy()); });
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
            _this._rings = [];
        }
        return _this;
    }
    Object.defineProperty(CurvePolygon.prototype, "rings", {
        /**
         * Get the rings
         * @return rings
         */
        get: function () {
            return this._rings;
        },
        /**
         * Set the rings
         * @param rings rings
         */
        set: function (rings) {
            var _this = this;
            this._rings = [];
            rings.forEach(function (ring) { return _this.addRing(ring); });
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add a ring
     * @param ring ring
     */
    CurvePolygon.prototype.addRing = function (ring) {
        this._rings.push(ring);
        this.updateZM(ring);
    };
    /**
     * Add rings
     * @param rings rings
     */
    CurvePolygon.prototype.addRings = function (rings) {
        var _this = this;
        rings.forEach(function (ring) { return _this.addRing(ring); });
    };
    /**
     * Get the number of rings including exterior and interior
     * @return number of rings
     */
    CurvePolygon.prototype.numRings = function () {
        return this._rings.length;
    };
    /**
     * Returns the Nth ring where the exterior ring is at 0, interior rings
     * begin at 1
     * @param n nth ring to return
     * @return ring
     */
    CurvePolygon.prototype.getRing = function (n) {
        return this._rings[n];
    };
    /**
     * Get the exterior ring
     *
     * @return exterior ring
     */
    CurvePolygon.prototype.getExteriorRing = function () {
        return this._rings[0];
    };
    /**
     * Get the number of interior rings
     *
     * @return number of interior rings
     */
    CurvePolygon.prototype.numInteriorRings = function () {
        return this._rings.length - 1;
    };
    /**
     * Returns the Nth interior ring for this Polygon
     * @param n interior ring number
     * @return interior ring
     */
    CurvePolygon.prototype.getInteriorRing = function (n) {
        return this._rings[n + 1];
    };
    /**
     * {@inheritDoc}
     */
    CurvePolygon.prototype.copy = function () {
        return new CurvePolygon(this);
    };
    /**
     * {@inheritDoc}
     */
    CurvePolygon.prototype.isEmpty = function () {
        return this._rings.length === 0;
    };
    /**
     * {@inheritDoc}
     */
    CurvePolygon.prototype.isSimple = function () {
        throw new internal_1.UnsupportedOperationException("Is Simple not implemented for CurvePolygon");
    };
    /**
     * {@inheritDoc}
     */
    CurvePolygon.prototype.equals = function (obj) {
        var equal = true;
        if (obj instanceof CurvePolygon) {
            if (this.numRings() === obj.numRings()) {
                for (var i = 0; i < this.numRings(); i++) {
                    if (!this.getRing(i).equals(obj.getRing(i))) {
                        equal = false;
                        break;
                    }
                }
            }
            else {
                equal = false;
            }
        }
        else {
            equal = false;
        }
        return equal;
    };
    return CurvePolygon;
}(internal_1.Surface));
exports.CurvePolygon = CurvePolygon;
