// @ngageoint/simple-features-js/dist/lib/GeometryType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeometryType = void 0;
/**
 * Geometry Type enumeration
 */
var GeometryType;
(function (GeometryType) {
    /**
     * The root of the geometry type hierarchy
     */
    GeometryType[GeometryType["GEOMETRY"] = 0] = "GEOMETRY";
    /**
     * A single location in space. Each point has an X and Y coordinate. A point
     * MAY optionally also have a Z and/or an M value.
     */
    GeometryType[GeometryType["POINT"] = 1] = "POINT";
    /**
     * A Curve that connects two or more points in space.
     */
    GeometryType[GeometryType["LINESTRING"] = 2] = "LINESTRING";
    /**
     * A restricted form of CurvePolygon where each ring is defined as a simple,
     * closed LineString.
     */
    GeometryType[GeometryType["POLYGON"] = 3] = "POLYGON";
    /**
     * A restricted form of GeometryCollection where each Geometry in the
     * collection must be of type Point.
     */
    GeometryType[GeometryType["MULTIPOINT"] = 4] = "MULTIPOINT";
    /**
     * A restricted form of MultiCurve where each Curve in the collection must
     * be of type LineString.
     */
    GeometryType[GeometryType["MULTILINESTRING"] = 5] = "MULTILINESTRING";
    /**
     * A restricted form of MultiSurface where each Surface in the collection
     * must be of type Polygon.
     */
    GeometryType[GeometryType["MULTIPOLYGON"] = 6] = "MULTIPOLYGON";
    /**
     * A collection of zero or more Geometry instances.
     */
    GeometryType[GeometryType["GEOMETRYCOLLECTION"] = 7] = "GEOMETRYCOLLECTION";
    /**
     * Circular String, Curve sub type
     */
    GeometryType[GeometryType["CIRCULARSTRING"] = 8] = "CIRCULARSTRING";
    /**
     * Compound Curve, Curve sub type
     */
    GeometryType[GeometryType["COMPOUNDCURVE"] = 9] = "COMPOUNDCURVE";
    /**
     * A planar surface defined by an exterior ring and zero or more interior
     * ring. Each ring is defined by a Curve instance.
     */
    GeometryType[GeometryType["CURVEPOLYGON"] = 10] = "CURVEPOLYGON";
    /**
     * A restricted form of GeometryCollection where each Geometry in the
     * collection must be of type Curve.
     */
    GeometryType[GeometryType["MULTICURVE"] = 11] = "MULTICURVE";
    /**
     * A restricted form of GeometryCollection where each Geometry in the
     * collection must be of type Surface.
     */
    GeometryType[GeometryType["MULTISURFACE"] = 12] = "MULTISURFACE";
    /**
     * The base type for all 1-dimensional geometry types. A 1-dimensional
     * geometry is a geometry that has a length, but no area. A curve is
     * considered simple if it does not intersect itself (except at the start
     * and end point). A curve is considered closed its start and end point are
     * coincident. A simple, closed curve is called a ring.
     */
    GeometryType[GeometryType["CURVE"] = 13] = "CURVE";
    /**
     * The base type for all 2-dimensional geometry types. A 2-dimensional
     * geometry is a geometry that has an area.
     */
    GeometryType[GeometryType["SURFACE"] = 14] = "SURFACE";
    /**
     * Contiguous collection of polygons which share common boundary segments.
     */
    GeometryType[GeometryType["POLYHEDRALSURFACE"] = 15] = "POLYHEDRALSURFACE";
    /**
     * A tetrahedron (4 triangular faces), corner at the origin and each unit
     * coordinate digit.
     */
    GeometryType[GeometryType["TIN"] = 16] = "TIN";
    /**
     * Triangle
     */
    GeometryType[GeometryType["TRIANGLE"] = 17] = "TRIANGLE";
})(GeometryType = exports.GeometryType || (exports.GeometryType = {}));
// eslint-disable-next-line @typescript-eslint/no-namespace
(function (GeometryType) {
    function nameFromType(type) {
        var name = null;
        if (type !== null && type !== undefined) {
            name = GeometryType[type];
        }
        return name;
    }
    GeometryType.nameFromType = nameFromType;
    function fromName(type) {
        return GeometryType[type];
    }
    GeometryType.fromName = fromName;
    function values() {
        return [
            GeometryType.GEOMETRY,
            GeometryType.POINT,
            GeometryType.LINESTRING,
            GeometryType.POLYGON,
            GeometryType.MULTIPOINT,
            GeometryType.MULTILINESTRING,
            GeometryType.MULTIPOLYGON,
            GeometryType.GEOMETRYCOLLECTION,
            GeometryType.CIRCULARSTRING,
            GeometryType.COMPOUNDCURVE,
            GeometryType.CURVEPOLYGON,
            GeometryType.MULTICURVE,
            GeometryType.MULTISURFACE,
            GeometryType.CURVE,
            GeometryType.SURFACE,
            GeometryType.POLYHEDRALSURFACE,
            GeometryType.TIN,
            GeometryType.TRIANGLE
        ];
    }
    GeometryType.values = values;
})(GeometryType = exports.GeometryType || (exports.GeometryType = {}));
