// @ngageoint/simple-features-js/dist/lib/LineString.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineString = void 0;
var internal_1 = require("./internal");
/**
 * A Curve that connects two or more points in space.
 */
var LineString = /** @class */ (function (_super) {
    __extends(LineString, _super);
    /**
     * Constructor
     */
    function LineString() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.LINESTRING, false, false) || this;
            _this._points = [];
        }
        else if (args.length === 1 && args[0] instanceof LineString) {
            _this = _super.call(this, internal_1.GeometryType.LINESTRING, args[0].hasZ, args[0].hasM) || this;
            _this._points = [];
            args[0].points.forEach(function (point) { return _this.addPoint(point.copy()); });
        }
        else if (args.length === 1 && args[0] instanceof Array) {
            _this = _super.call(this, internal_1.GeometryType.LINESTRING, internal_1.Geometry.hasZ(args[0]), internal_1.Geometry.hasM(args[0])) || this;
            _this._points = args[0];
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.LINESTRING, args[0], args[1]) || this;
            _this._points = [];
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
            _this._points = [];
        }
        return _this;
    }
    Object.defineProperty(LineString.prototype, "points", {
        /**
         * Get the points
         * @return points
         */
        get: function () {
            return this._points;
        },
        /**
         * Set the points
         * @param points points
         */
        set: function (points) {
            this._points = points;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add a point
     * @param point point
     */
    LineString.prototype.addPoint = function (point) {
        this._points.push(point);
        this.updateZM(point);
    };
    /**
     * Add points
     * @param points points
     */
    LineString.prototype.addPoints = function (points) {
        var _this = this;
        points.forEach(function (point) { return _this.addPoint(point); });
    };
    /**
     * Get the number of points
     * @return number of points
     */
    LineString.prototype.numPoints = function () {
        return this._points.length;
    };
    /**
     * Returns the Nth point
     * @param n nth point to return
     * @return point
     */
    LineString.prototype.getPoint = function (n) {
        return this._points[n];
    };
    /**
     * {@inheritDoc}
     */
    LineString.prototype.startPoint = function () {
        var startPoint = null;
        if (!this.isEmpty()) {
            startPoint = this._points[0];
        }
        return startPoint;
    };
    /**
     * {@inheritDoc}
     */
    LineString.prototype.endPoint = function () {
        var endPoint = null;
        if (!this.isEmpty()) {
            endPoint = this._points[this._points.length - 1];
        }
        return endPoint;
    };
    /**
     * {@inheritDoc}
     */
    LineString.prototype.isSimple = function () {
        return internal_1.ShamosHoey.simplePolygonPoints(this._points);
    };
    /**
     * {@inheritDoc}
     */
    LineString.prototype.copy = function () {
        return new LineString(this);
    };
    /**
     * {@inheritDoc}
     */
    LineString.prototype.isEmpty = function () {
        return this._points.length === 0;
    };
    /**
     * {@inheritDoc}
     */
    LineString.prototype.equals = function (obj) {
        var equal = true;
        if (obj instanceof LineString && this.numPoints() === obj.numPoints()) {
            for (var i = 0; i < this._points.length; i++) {
                if (!this.getPoint(i).equals(obj.getPoint(i))) {
                    equal = false;
                    break;
                }
            }
        }
        else {
            equal = false;
        }
        return equal;
    };
    return LineString;
}(internal_1.Curve));
exports.LineString = LineString;
