// @ngageoint/simple-features-js/dist/lib/MultiPolygon.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiPolygon = void 0;
var internal_1 = require("./internal");
/**
 * A restricted form of MultiSurface where each Surface in the collection must
 * be of type Polygon.
 */
var MultiPolygon = /** @class */ (function (_super) {
    __extends(MultiPolygon, _super);
    /**
     * Constructor
     */
    function MultiPolygon() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOLYGON, false, false) || this;
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOLYGON, args[0], args[1]) || this;
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
        }
        else if (args.length === 1 && args[0] instanceof internal_1.Polygon) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOLYGON, args[0].hasZ, args[0].hasM) || this;
            _this.addPolygon(args[0]);
        }
        else if (args.length === 1 && args[0] instanceof MultiPolygon) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOLYGON, args[0].hasZ, args[0].hasM) || this;
            args[0].polygons.forEach(function (polygon) { return _this.addPolygon(polygon); });
        }
        else if (args.length === 1 && args[0].length != null) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOLYGON, internal_1.Geometry.hasZ(args[0]), internal_1.Geometry.hasM(args[0])) || this;
            _this.polygons = args[0];
        }
        return _this;
    }
    Object.defineProperty(MultiPolygon.prototype, "polygons", {
        /**
         * Get the polygons
         * @return polygons
         */
        get: function () {
            return this.getSurfaces();
        },
        /**
         * Set the polygons
         * @param polygons polygons
         */
        set: function (polygons) {
            this.setSurfaces(polygons);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add a polygon
     * @param polygon polygon
     */
    MultiPolygon.prototype.addPolygon = function (polygon) {
        this.addSurface(polygon);
    };
    /**
     * Add polygons
     * @param polygons polygons
     */
    MultiPolygon.prototype.addPolygons = function (polygons) {
        this.addSurfaces(polygons);
    };
    /**
     * Get the number of polygons
     * @return number of polygons
     */
    MultiPolygon.prototype.numPolygons = function () {
        return this.numSurfaces();
    };
    /**
     * Returns the Nth polygon
     * @param n  nth polygon to return
     * @return polygon
     */
    MultiPolygon.prototype.getPolygon = function (n) {
        return this.getSurface(n);
    };
    /**
     * {@inheritDoc}
     */
    MultiPolygon.prototype.copy = function () {
        return new MultiPolygon(this);
    };
    return MultiPolygon;
}(internal_1.MultiSurface));
exports.MultiPolygon = MultiPolygon;
