// @ngageoint/simple-features-js/dist/lib/MultiLineString.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiLineString = void 0;
var internal_1 = require("./internal");
/**
 * A restricted form of MultiCurve where each Curve in the collection must be of
 * type LineString.
 */
var MultiLineString = /** @class */ (function (_super) {
    __extends(MultiLineString, _super);
    /**
     * Constructor
     */
    function MultiLineString() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.MULTILINESTRING, false, false) || this;
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.MULTILINESTRING, args[0], args[1]) || this;
        }
        else if (args.length === 1 && args[0] instanceof internal_1.LineString) {
            _this = _super.call(this, internal_1.GeometryType.MULTILINESTRING, args[0].hasZ, args[0].hasM) || this;
            _this.addLineString(args[0]);
        }
        else if (args.length === 1 && args[0] instanceof MultiLineString) {
            _this = _super.call(this, internal_1.GeometryType.MULTILINESTRING, args[0].hasZ, args[0].hasM) || this;
            args[0].lineStrings.forEach(function (lineString) { return _this.addLineString(lineString); });
        }
        else if (args.length === 1 && args[0].length != null) {
            _this = _super.call(this, internal_1.GeometryType.MULTILINESTRING, internal_1.GeometryCollection.hasZ(args[0]), internal_1.GeometryCollection.hasM(args[0])) || this;
            _this.lineStrings = args[0];
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
        }
        return _this;
    }
    Object.defineProperty(MultiLineString.prototype, "lineStrings", {
        /**
         * Get the line strings
         * @return line strings
         */
        get: function () {
            return this.getCurves();
        },
        /**
         * Set the line strings
         *
         * @param lineStrings
         *            line strings
         */
        set: function (lineStrings) {
            this.setCurves(lineStrings);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add a line string
     *
     * @param lineString
     *            line string
     */
    MultiLineString.prototype.addLineString = function (lineString) {
        this.addCurve(lineString);
    };
    /**
     * Add line strings
     *
     * @param lineStrings
     *            line strings
     */
    MultiLineString.prototype.addLineStrings = function (lineStrings) {
        this.addCurves(lineStrings);
    };
    /**
     * Get the number of line strings
     *
     * @return number of line strings
     */
    MultiLineString.prototype.numLineStrings = function () {
        return this.numCurves();
    };
    /**
     * Returns the Nth line string
     * @param n nth line string to return
     * @return line string
     */
    MultiLineString.prototype.getLineString = function (n) {
        return this.getCurve(n);
    };
    /**
     * {@inheritDoc}
     */
    MultiLineString.prototype.copy = function () {
        return new MultiLineString(this);
    };
    return MultiLineString;
}(internal_1.MultiCurve));
exports.MultiLineString = MultiLineString;
