// @ngageoint/simple-features-js/dist/lib/MultiCurve.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiCurve = void 0;
var internal_1 = require("./internal");
/**
 * A restricted form of GeometryCollection where each Geometry in the collection
 * must be of type Curve.
 * @param <T> curve type
 */
var MultiCurve = /** @class */ (function (_super) {
    __extends(MultiCurve, _super);
    /**
     * Constructor
     * @param type geometry type
     * @param hasZ has z
     * @param hasM has m
     */
    function MultiCurve(type, hasZ, hasM) {
        return _super.call(this, type, hasZ, hasM) || this;
    }
    /**
     * Get the curves
     * @return curves
     */
    MultiCurve.prototype.getCurves = function () {
        return this.geometries;
    };
    /**
     * Set the curves
     * @param curves curves
     */
    MultiCurve.prototype.setCurves = function (curves) {
        this.geometries = curves;
    };
    /**
     * Add a curve
     * @param curve curve
     */
    MultiCurve.prototype.addCurve = function (curve) {
        this.addGeometry(curve);
    };
    /**
     * Add curves
     * @param curves curves
     */
    MultiCurve.prototype.addCurves = function (curves) {
        this.addGeometries(curves);
    };
    /**
     * Get the number of curves
     * @return number of curves
     */
    MultiCurve.prototype.numCurves = function () {
        return this.numGeometries();
    };
    /**
     * Returns the Nth curve
     * @param n nth line curve to return
     * @return curve
     */
    MultiCurve.prototype.getCurve = function (n) {
        return this.getGeometry(n);
    };
    /**
     * Determine if this Multi Curve is closed for each Curve (start point = end
     * point)
     *
     * @return true if closed
     */
    MultiCurve.prototype.isClosed = function () {
        var e_1, _a;
        var closed = true;
        try {
            for (var _b = __values(this.geometries), _c = _b.next(); !_c.done; _c = _b.next()) {
                var curve = _c.value;
                if (!curve.isClosed()) {
                    closed = false;
                    break;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return closed;
    };
    return MultiCurve;
}(internal_1.GeometryCollection));
exports.MultiCurve = MultiCurve;
