// @ngageoint/simple-features-js/dist/lib/MultiPoint.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiPoint = void 0;
var internal_1 = require("./internal");
/**
 * A restricted form of GeometryCollection where each Geometry in the collection
 * must be of type Point.
 */
var MultiPoint = /** @class */ (function (_super) {
    __extends(MultiPoint, _super);
    /**
     * Constructor
     */
    function MultiPoint() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOINT, false, false) || this;
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOINT, args[0], args[1]) || this;
        }
        else if (args.length === 1 && args[0] instanceof internal_1.Point) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOINT, args[0].hasZ, args[0].hasM) || this;
            _this.addPoint(args[0]);
        }
        else if (args.length === 1 && args[0] instanceof MultiPoint) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOINT, args[0].hasZ, args[0].hasM) || this;
            args[0].points.forEach(function (point) { return _this.addPoint(point); });
        }
        else if (args.length === 1 && args[0].length != null) {
            _this = _super.call(this, internal_1.GeometryType.MULTIPOINT, internal_1.GeometryCollection.hasZ(args[0]), internal_1.GeometryCollection.hasM(args[0])) || this;
            _this.points = args[0];
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
        }
        return _this;
    }
    Object.defineProperty(MultiPoint.prototype, "points", {
        /**
         * Get the points
         * @return points
         */
        get: function () {
            return this.geometries;
        },
        /**
         * Set the points
         * @param points points
         */
        set: function (points) {
            this.geometries = points;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add a point
     * @param point point
     */
    MultiPoint.prototype.addPoint = function (point) {
        this.addGeometry(point);
    };
    /**
     * Add points
     * @param points points
     */
    MultiPoint.prototype.addPoints = function (points) {
        this.addGeometries(points);
    };
    /**
     * Get the number of points
     * @return number of points
     */
    MultiPoint.prototype.numPoints = function () {
        return this.numGeometries();
    };
    /**
     * Returns the Nth point
     *
     * @param n nth point to return
     * @return point
     */
    MultiPoint.prototype.getPoint = function (n) {
        return this.getGeometry(n);
    };
    /**
     * {@inheritDoc}
     */
    MultiPoint.prototype.copy = function () {
        return new MultiPoint(this);
    };
    /**
     * {@inheritDoc}
     */
    MultiPoint.prototype.isSimple = function () {
        var _this = this;
        return this.points.filter(function (pA, index) {
            return _this.points.findIndex(function (pB) { return pA.equals(pB); }) === index;
        }).length == this.numPoints();
    };
    return MultiPoint;
}(internal_1.GeometryCollection));
exports.MultiPoint = MultiPoint;
