// @ngageoint/simple-features-js/dist/lib/MultiSurface.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiSurface = void 0;
var internal_1 = require("./internal");
/**
 * A restricted form of GeometryCollection where each Geometry in the collection
 * must be of type Surface.
 *
 * @param <T> surface type
 */
var MultiSurface = /** @class */ (function (_super) {
    __extends(MultiSurface, _super);
    /**
     * Constructor
     *
     * @param type geometry type
     * @param hasZ has z
     * @param hasM has m
     */
    function MultiSurface(type, hasZ, hasM) {
        return _super.call(this, type, hasZ, hasM) || this;
    }
    /**
     * Get the surfaces
     * @return surfaces
     */
    MultiSurface.prototype.getSurfaces = function () {
        return this.geometries;
    };
    /**
     * Set the surfaces
     * @param surfaces surfaces
     */
    MultiSurface.prototype.setSurfaces = function (surfaces) {
        this.geometries = surfaces;
    };
    /**
     * Add a surface
     * @param surface surface
     */
    MultiSurface.prototype.addSurface = function (surface) {
        this.addGeometry(surface);
    };
    /**
     * Add surfaces
     *
     * @param surfaces
     *            surfaces
     */
    MultiSurface.prototype.addSurfaces = function (surfaces) {
        this.addGeometries(surfaces);
    };
    /**
     * Get the number of surfaces
     * @return number of surfaces
     */
    MultiSurface.prototype.numSurfaces = function () {
        return this.numGeometries();
    };
    /**
     * Returns the Nth surface
     * @param n nth line surface to return
     * @return surface
     */
    MultiSurface.prototype.getSurface = function (n) {
        return this.getGeometry(n);
    };
    return MultiSurface;
}(internal_1.GeometryCollection));
exports.MultiSurface = MultiSurface;
