// @ngageoint/simple-features-js/dist/lib/Point.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Point = void 0;
var internal_1 = require("./internal");
/**
 * A single location in space. Each point has an X and Y coordinate. A point MAY
 * optionally also have a Z and/or an M value.
 */
var Point = /** @class */ (function (_super) {
    __extends(Point, _super);
    /**
     * Constructor
     */
    function Point() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.POINT, false, false) || this;
            _this._x = 0;
            _this._y = 0;
        }
        else if (args.length === 1 && args[0] instanceof Point) {
            _this = _super.call(this, internal_1.GeometryType.POINT, args[0].hasZ, args[0].hasM) || this;
            _this._x = args[0].x;
            _this._y = args[0].y;
            _this._z = args[0].z;
            _this._m = args[0].m;
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.POINT, false, false) || this;
            _this._x = args[0];
            _this._y = args[1];
        }
        else if (args.length === 3) {
            _this = _super.call(this, internal_1.GeometryType.POINT, args[2] != null, false) || this;
            _this._x = args[0];
            _this._y = args[1];
            _this._z = args[2];
        }
        else if (args.length === 4) {
            if (typeof args[0] === "boolean" && typeof args[1] === "boolean") {
                _this = _super.call(this, internal_1.GeometryType.POINT, args[0], args[1]) || this;
                _this._x = args[2];
                _this._y = args[3];
            }
            else {
                _this = _super.call(this, internal_1.GeometryType.POINT, args[2] != null, args[3] != null) || this;
                _this._x = args[0];
                _this._y = args[1];
                _this._z = args[2];
                _this._m = args[3];
            }
        }
        return _this;
    }
    Object.defineProperty(Point.prototype, "x", {
        /**
         * Get x
         * @return x
         */
        get: function () {
            return this._x;
        },
        /**
         * Set y
         * @param x
         */
        set: function (x) {
            this._x = x;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Point.prototype, "y", {
        /**
         * Get y
         * @return y
         */
        get: function () {
            return this._y;
        },
        /**
         * Set y
         * @param y
         */
        set: function (y) {
            this._y = y;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Point.prototype, "z", {
        /**
         * Get z
         * @return z
         */
        get: function () {
            return this._z;
        },
        /**
         * Set z
         * @param z
         */
        set: function (z) {
            this._z = z;
            this.hasZ = z != null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Point.prototype, "m", {
        /**
         * Get m
         *
         * @return m
         */
        get: function () {
            return this._m;
        },
        /**
         * Set m
         * @param m
         */
        set: function (m) {
            this._m = m;
            this.hasM = m != null;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * {@inheritDoc}
     */
    Point.prototype.copy = function () {
        return new Point(this);
    };
    /**
     * {@inheritDoc}
     */
    Point.prototype.isEmpty = function () {
        return false;
    };
    /**
     * {@inheritDoc}
     */
    Point.prototype.isSimple = function () {
        return true;
    };
    /**
     * {@inheritDoc}
     */
    Point.prototype.equals = function (obj) {
        return _super.prototype.equals.call(this, obj) && obj instanceof Point && this.m === obj.m && this.z === obj.z && this.x === obj.x && this.y === obj.y;
    };
    return Point;
}(internal_1.Geometry));
exports.Point = Point;
