// @ngageoint/simple-features-js/dist/lib/PolyhedralSurface.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolyhedralSurface = void 0;
var internal_1 = require("./internal");
/**
 * Contiguous collection of polygons which share common boundary segments.
 *
 * @author osbornb
 */
var PolyhedralSurface = /** @class */ (function (_super) {
    __extends(PolyhedralSurface, _super);
    /**
     * Constructor
     */
    function PolyhedralSurface() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = this;
        if (args.length === 0) {
            _this = _super.call(this, internal_1.GeometryType.POLYHEDRALSURFACE, false, false) || this;
            _this._polygons = [];
        }
        else if (args.length === 1 && args[0] instanceof internal_1.Polygon) {
            _this = _super.call(this, internal_1.GeometryType.POLYHEDRALSURFACE, args[0].hasZ, args[0].hasM) || this;
            _this._polygons = [];
            _this.addPolygon(args[0]);
        }
        else if (args.length === 1 && args[0] instanceof PolyhedralSurface) {
            _this = _super.call(this, internal_1.GeometryType.POLYHEDRALSURFACE, args[0].hasZ, args[0].hasM) || this;
            _this._polygons = [];
            args[0].polygons.forEach(function (polygon) { return _this.addPolygon(polygon.copy()); });
        }
        else if (args.length === 1 && args[0].length != null) {
            _this = _super.call(this, internal_1.GeometryType.POLYHEDRALSURFACE, args[0].hasZ, args[0].hasM) || this;
            _this._polygons = args[0];
        }
        else if (args.length === 2) {
            _this = _super.call(this, internal_1.GeometryType.POLYHEDRALSURFACE, args[0], args[1]) || this;
            _this._polygons = [];
        }
        else if (args.length === 3) {
            _this = _super.call(this, args[0], args[1], args[2]) || this;
            _this._polygons = [];
        }
        return _this;
    }
    Object.defineProperty(PolyhedralSurface.prototype, "polygons", {
        /**
         * Get polygons
         * @return polygons
         */
        get: function () {
            return this._polygons;
        },
        /**
         * Set polygons
         *
         * @param polygons
         *            polygons
         */
        set: function (polygons) {
            this._polygons = polygons;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolyhedralSurface.prototype, "patches", {
        /**
         * Get patches
         *
         * @return patches
         * @see #getPolygons()
         */
        get: function () {
            return this._polygons;
        },
        /**
         * Set patches
         *
         * @param patches
         *            patches
         * @see #setPolygons(List)
         */
        set: function (polygons) {
            this._polygons = polygons;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Add polygon
     * @param polygon polygon
     */
    PolyhedralSurface.prototype.addPolygon = function (polygon) {
        this._polygons.push(polygon);
        this.updateZM(polygon);
    };
    /**
     * Add patch
     * @param patch patch
     * @see #addPolygon(Polygon)
     */
    PolyhedralSurface.prototype.addPatch = function (patch) {
        this.addPolygon(patch);
    };
    /**
     * Add polygons
     * @param polygons polygons
     */
    PolyhedralSurface.prototype.addPolygons = function (polygons) {
        var e_1, _a;
        try {
            for (var polygons_1 = __values(polygons), polygons_1_1 = polygons_1.next(); !polygons_1_1.done; polygons_1_1 = polygons_1.next()) {
                var polygon = polygons_1_1.value;
                this.addPolygon(polygon);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (polygons_1_1 && !polygons_1_1.done && (_a = polygons_1.return)) _a.call(polygons_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * Add patches
     * @param patches patches
     * @see #addPolygons(List)
     */
    PolyhedralSurface.prototype.addPatches = function (patches) {
        this.addPolygons(patches);
    };
    /**
     * Get the number of polygons
     * @return number of polygons
     */
    PolyhedralSurface.prototype.numPolygons = function () {
        return this._polygons.length;
    };
    /**
     * Get the number of polygons
     * @return number of polygons
     * @see #numPolygons()
     */
    PolyhedralSurface.prototype.numPatches = function () {
        return this.numPolygons();
    };
    /**
     * Get the Nth polygon
     * @param n nth polygon to return
     * @return polygon
     */
    PolyhedralSurface.prototype.getPolygon = function (n) {
        return this._polygons[n];
    };
    /**
     * Get the Nth polygon patch
     * @param n nth polygon patch to return
     * @return polygon patch
     * @see #getPolygon(int)
     */
    PolyhedralSurface.prototype.getPatch = function (n) {
        return this.getPolygon(n);
    };
    /**
     * {@inheritDoc}
     */
    PolyhedralSurface.prototype.copy = function () {
        return new PolyhedralSurface(this);
    };
    /**
     * {@inheritDoc}
     */
    PolyhedralSurface.prototype.isEmpty = function () {
        return this._polygons.length === 0;
    };
    /**
     * {@inheritDoc}
     */
    PolyhedralSurface.prototype.isSimple = function () {
        throw new internal_1.UnsupportedOperationException("Is Simple not implemented for PolyhedralSurface");
    };
    /**
     * {@inheritDoc}
     */
    PolyhedralSurface.prototype.equals = function (obj) {
        var equal = true;
        if (obj instanceof PolyhedralSurface && this.numPatches() === obj.numPatches()) {
            for (var i = 0; i < this.numPatches(); i++) {
                if (!this.getPatch(i).equals(obj.getPatch(i))) {
                    equal = false;
                    break;
                }
            }
        }
        else {
            equal = false;
        }
        return equal;
    };
    return PolyhedralSurface;
}(internal_1.Surface));
exports.PolyhedralSurface = PolyhedralSurface;
